test/Cimpress.Extensions.Http.UnitTests/MessageHandlers/ConcurrencyLimitingHandler_when_limiting_concurrency.cs

```diff
@@ -0,0 +1,89 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net;
+using System.Net.Http;
+using System.Threading;
+using System.Threading.Tasks;
+using Cimpress.Extensions.Http.MessageHandlers;
+using FluentAssertions;
+using Xunit;
+
+namespace Cimpress.Extensions.Http.UnitTests.MessageHandlers
+{
+    public class ConcurrencyLimitingHandler_when_limiting_concurrency
+    {
+        private class TestHandler : DelegatingHandler
+        {
+            private readonly TimeSpan executionTime = TimeSpan.FromMilliseconds(10);
+            private int numberOfExecutions;
+            private readonly int maxNumberOfExecutions;
+            
+            public int MaxSeenNumberOfExecutions { get; private set; }
+
+            public TestHandler(int maxNumberOfExecutions)
+            {
+                this.maxNumberOfExecutions = maxNumberOfExecutions;
+            }
+            
+            protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+            {
+                numberOfExecutions++;
+                MaxSeenNumberOfExecutions = Math.Max(numberOfExecutions, MaxSeenNumberOfExecutions);
+                if (numberOfExecutions > maxNumberOfExecutions)
+                {
+                    throw new Exception("Max number of execution reached.");
+                }
+                await Task.Delay(executionTime, cancellationToken);
+                var msg = new HttpResponseMessage(HttpStatusCode.OK);
+                numberOfExecutions--;
+                return msg;
+            }
+        }
+
+        public static IEnumerable<object[]> TestData
+        {
+            get
+            {
+                var urls = Enumerable.Range(0, 100).Select(r => new Uri($"http://unit.test/{r}"));
+                yield return new object[] {urls, 10};
+                yield return new object[] {urls, 5};
+                yield return new object[] {urls, 20};
+            }
+        }
+
+        [Theory]
+        [MemberData(nameof(TestData))]
+        public void Does_not_exceed_concurrency(IEnumerable<Uri> uris, int maxConcurrency)
```

Comment:
We could have done this in a non-time dependent way right?

Hmmm, but looking at it there isn't a great way to capture that, without creating some sort of hacky testHandler to capture the requests coming in, and then using those to drive the firing cancellation tokens which would cancel the active ones.